distributed-prover/src/circuits/evm.rs

```diff
@@ -0,0 +1,216 @@
+    util::SubCircuit
+};
+
+use crate::util::artifacts::*;
+use crate::util::constants::RNG_SEED;
+
+pub(crate) const CIRCUIT_NAME: &str = "evm";
+const CIRCUIT_DEGREE: u32 = 18;
+
+// What is the <Fr> notation?
+// What is the pub(crate) notation?
+pub(crate) fn circuit() -> EvmCircuit<Fr> { 
+    let timer = start_timer!(|| "Create circuit");  // What is the "||" Notation?
+
+    let empty_data: GethData = TestContext::<0, 0>::new(None, |_| {}, |_, _| {}, |b, _| b)
+        .unwrap()
+        .into();
+
+    let mut builder =
+        BlockData::new_from_geth_data_with_params(empty_data.clone(), FixedCParams::default())
+            .new_circuit_input_builder();
+
+    builder
+        .handle_block(&empty_data.eth_block, &empty_data.geth_traces)
+        .unwrap();
+
+    let block = block_convert(&builder).unwrap();
+
+    let circuit = TestEvmCircuit::<Fr>::new(block);
+
+    end_timer!(timer);
+
+    circuit
+}
+
+pub(crate) fn setup() {
+    let circuit = circuit();
+
+    // Setup params
+    let timer = start_timer!(|| "Set up params");
+    let mut rng = XorShiftRng::from_seed(RNG_SEED);
+    let general_params = ParamsKZG::<Bn256>::setup(CIRCUIT_DEGREE, &mut rng);
+    let verifier_params = general_params.verifier_params().clone();     // What's the purpose of cloning here?
+    end_timer!(timer);
+
+    // Generate Verification Key
+    let timer = start_timer!(|| "Generate Verification Key");
+    let vk = keygen_vk(&general_params, &circuit).unwrap();
+    end_timer!(timer);
+
+    // Read Network Configuration
+    let timer = start_timer!(|| "Read Network Configuration");
+    let network_config = read_workload_config(CIRCUIT_NAME);
+    end_timer!(timer);
+    let num_prover = network_config.num_prover();
+
+    // Generate Workload Configuration
+    let timer = start_timer!(|| "Generate Workload Configuration");
+    // What is in a worklaod config? What is G1Affine? Is it an algorithm?
+    let workload_config = WorkloadConfig::new_even_distribution::<G1Affine>(vk.cs(), num_prover);
+    end_timer!(timer);
+
+    // Generate Proving Key
+    let timer = start_timer!(|| "Generate Proving Key");
+    // Why we need to clone the vk to generate pk? Why can't we pass &vk to make it immutable
+    let pk = keygen_pk(&general_params, vk.clone(), &circuit).unwrap();
+    end_timer!(timer);
+
+    // Artifact Serialization
+    let timer = start_timer!(|| "Artifact Serialization");
+    write_params_kzg(CIRCUIT_DEGREE, &general_params, false);
+    write_params_kzg(CIRCUIT_DEGREE, &verifier_params, true);
+    write_vk(CIRCUIT_NAME, &vk);
+    write_workload_config(CIRCUIT_NAME, &workload_config);
+    write_pk(CIRCUIT_NAME, &pk);
+    end_timer!(timer);
+}
+
+pub(crate) fn prove(prover_index: usize) {
+    let rng = XorShiftRng::from_seed(RNG_SEED);
+    let circuit = circuit();
+    // What is special about blake2b here and why are we using this instead of other hashing?
+    // What is <_, G1Affine, Challenge255<_>> in between ::s?
```

Comment:
Similar to <Fr> before. Here, a generic function (not generic struct) needs to be given specific type parameters.